### DISCUSSION TITLE
Help on understanding error: "RuntimeError: The following cell name(s) are used for more than one cell - can't write this layout"

### DISCUSSION THREAD

**Discussion by duarte-jfs** [gdsfactory/gdsfactory#2949] [Q&A]
I have a complicated cell that generates the following gds:

![image](https://github.com/gdsfactory/gdsfactory/assets/60704674/1fd5a668-b3cb-4aeb-937a-f56a4febecbd)

There are two functions to generate this. The first is called `spiral_array_loopback()` and it runs just fine. Then I created a second function, it every way similar to the previous one called `spiral_array_loopback_N2_WG1()`. This second one differs from the first purely by two custom tapers. On the second one though, I get the error:

> RuntimeError: The following cell name(s) are used for more than one cell - can't write this layout:
> taper_W2_W0p1_TL150_CSF_ffc61053 in Cell.write

To understand this I run the snippet:

```python

c = spiral_array_loopback()
    # c = spiral_array_loopback_N2_WG1()
    for i in c.insts:
        print(i.cell.name)
        for k in i.cell.insts:
            print('-', k.cell.name)
            
```

This returns (I will just print the two first spirals:

> spiral_loopback_WW2_TW3_4a7ef04c
> - spiral_L0_Bbend_euler_S_92b2f174
> - bend_euler_R100_A90_P0p_90a13251
> - bend_euler_R100_A90_P0p_90a13251
> - straight_L85_N2_CSxs_542340f0
> - straight_L285_N2_CSxs_542340f0
> - straight_L555_N2_CSxs_542340f0
> - straight_L4555_N2_CSxs_542340f0
> - edge_coupler_left_IW7_W_91ab590e
> - edge_coupler_right_IW7__759c3db4
> text_T0__L17064p2_um_|__75a24336
> spiral_loopback_WW2_TW3_e73769f8
> - spiral_L50_Bbend_euler__dc7659d6
> - bend_euler_R100_A90_P0p_90a13251
> - bend_euler_R100_A90_P0p_90a13251
> - straight_L85_N2_CSxs_542340f0
> - straight_L335_N2_CSxs_542340f0
> - straight_L1280_N2_CSxs_542340f0
> - straight_L3780_N2_CSxs_542340f0
> - edge_coupler_left_IW7_W_91ab590e
> - edge_coupler_right_IW7__759c3db4

The different `spiral_loopback` have different names, but their inner cells name are repeated between the `spiral_loopback`. In other words, as an example, both `spiral_loopback_WW2_TW3_e73769f8` and `spiral_loopback_WW2_TW3_4a7ef04c` have `bend_euler_R100_A90_P0p_90a13251`. This gives no error. When I run the second function, however I do get the above error and the instance analysis gives:

> spiral_loopback_N2_WG1__69667510
> - spiral_L0_Bbend_euler_S_ef99a00d
> - bend_euler_R100_A90_P0p_3e8603a4
> - bend_euler_R100_A90_P0p_3e8603a4
> - straight_L85_N2_CSxs_e2cbe220
> - straight_L285_N2_CSxs_e2cbe220
> - straight_L405_N2_CSxs_e2cbe220
> - straight_L4405_N2_CSxs_e2cbe220
> - edge_coupler_left_IW7_W_91ab590e
> - edge_coupler_right_IW7__759c3db4
> - taper_W2_W0p1_TL150_CSF_ffc61053
> - taper_W2_W0p1_TL150_CSF_ffc61053
> text_T0__L17064p2_um_|__75a24336
> spiral_loopback_N2_WG1__a7ed4335
> - spiral_L50_Bbend_euler__2e3f8323
> - bend_euler_R100_A90_P0p_3e8603a4
> - bend_euler_R100_A90_P0p_3e8603a4
> - straight_L85_N2_CSxs_e2cbe220
> - straight_L335_N2_CSxs_e2cbe220
> - straight_L1130_N2_CSxs_e2cbe220
> - straight_L3630_N2_CSxs_e2cbe220
> - edge_coupler_left_IW7_W_91ab590e
> - edge_coupler_right_IW7__759c3db4
> - taper_W2_W0p1_TL150_CSF_ffc61053
> - taper_W2_W0p1_TL150_CSF_ffc61053

It apparently it runs into a problem with `taper_W2_W0p1_TL150_CSF_ffc61053` as per the error above, but I cannot understand why it does not give an error before. Can someone help?

**Comment by sebastian-goeldi** [comment]
This happens if something is not cached properly. Which taper are you using?

The cell name of instances is not really relevant. What is relevant is the content of `gf.kcl.kcells`

If cells are in there with the same name multiple times, this can cause problems (if the cells are to be written)

**Reply by duarte-jfs** [reply]
> This happens if something is not cached properly. Which taper are you using?

I am using a custom taper. In hindsight, it was unnecessary to have this custom one. I can revert to the standard gdsfactory one. But that shouldn't be a problem, should it?

This is how it is defined:

```python

@gf.cell
def taper(width1: float = 2,
          width2: float = 3,
          taper_length: float = 20,
          cross_section_function = None):
    
    c = Component('taper1')

    _taper_init_xs = cross_section_function(width = width1)
    _taper_final_xs = cross_section_function(width = width2)

    xs_transition = gf.path.transition(cross_section1=_taper_init_xs,
                                       cross_section2=_taper_final_xs,
                                       width_type='linear')
    
    #create the path onto which to extrude the transition
    path_transition = gf.path.straight(length = taper_length,
                                        npoints = 100)
    
    _taper = gf.path.extrude_transition(path_transition,
                                        xs_transition)
    
    taper = c<<_taper

    c.add_port(name = 'o1', port = taper.ports['o1'])
    c.add_port(name = 'o2', port = taper.ports['o2'])

    c.add_route_info(cross_section=xs_transition, length = taper_length)

    
    # c.flatten()
    return c

```

**Reply by sebastian-goeldi** [reply]
I think the problem might be related to the extrusion. But I am not entirely sure.

I would try whether the error still persists if the cell is flattened at the end. I am not super familiar with the gdsfactory extrusion routines.

@joamatab might be able to help more

**Reply by duarte-jfs** [reply]
I have tried with the flattened cell, but it doesn't work. I believe I have narrowed down the problem to the cross_section specification. When I run the code snippet below, it seems that whenever the `partial` function is used to define the cross section at the time of instantiation of the component, the component is not cached properly. 

```python
from gdsfactory.cross_section import strip, Section

cross_section_func = partial(
    strip,
    sections=(Section(width=6, 
                      layer="SLAB90", 
                      name="slab", 
                      simplify=50 * 1e-3),),
    radius=100,
    radius_min=10,
)


for i in range(4):
    _taper = taper(width1 = 2,
                width2 = 0.1,
                taper_length = 150,
                cross_section_function=partial(
                                        strip,
                                        sections=(Section(width=6, 
                                                        layer="SLAB90", 
                                                        name="slab", 
                                                        simplify=50 * 1e-3),),
                                        radius=100,
                                        radius_min=10,
                                    ))
    
    _taper2 = taper(width1 = 2,
                width2 = 0.1,
                taper_length = 150,
                cross_section_function=cross_section_func)
    

    _straight = gf.components.straight(
        length = 150,
        cross_section=partial(
                                        strip,
                                        sections=(Section(width=6, 
                                                        layer="SLAB90", 
                                                        name="slab", 
                                                        simplify=50 * 1e-3),),
                                        radius=100,
                                        radius_min=10,
                                    ))
    
    _straight2 = gf.components.straight(
        length = 150,
        cross_section=cross_section_func)
    
    if i==0:
        previous_id_taper = id(_taper)
        previous_id_taper2 = id(_taper2)
        previous_id_straight = id(_straight)
        previous_id_straight2 = id(_straight2)
        continue
    else:
        print(previous_id_taper == id(_taper),
              previous_id_taper2 == id(_taper2),
              previous_id_straight == id(_straight),
              previous_id_straight2 == id(_straight2))
        
        previous_id_taper = id(_taper)
        previous_id_taper2 = id(_taper2)
        previous_id_straight = id(_straight)
        previous_id_straight2 = id(_straight2)
```

This returns:

> False True False True
> False True False True
> False True False True

**Reply by JinchW** [reply]
In version gdsfactory 8.7.0,  `c.flatten()` seems works. But the problem still exists without `c.flatten()`.

Is it possible to avoid this problem without using `c.flatten()`?

This problem occurs when using custom cross_section (or maybe the `partial()`).

The following is the test code, I hope it can help solve this problem.
```
import gdsfactory as gf
from functools import partial

def straight_taper(
        width1_taper: float = 0.2,
        length_wg: float = 20):
    """For demo the RuntimeError: The following cell name(s) are used for more than one cell - can't write this layout"""
    c = gf.Component()
    ref1_xs = partial(gf.cross_section.cross_section,
                      layer=(11, 1))
    # ref1_xs = "strip"
    ref1 = c << gf.components.taper(width1=width1_taper,
                                    width2=1,
                                    length=50,
                                    cross_section=ref1_xs
                                    )
    
    ref2_xs = partial(gf.cross_section.cross_section,
                      layer=(12, 1))
    # ref2_xs = "strip"
    ref2 = c << gf.components.straight(length=length_wg,
                                       cross_section=ref2_xs
                                       )
    ref2.connect("o1",ref1.ports["o1"],
                 allow_width_mismatch = True,
                 allow_layer_mismatch = True,
                 allow_type_mismatch = True)
    # c.flatten()
    return c

c = gf.Component()
for ii in range(3):    
    ref = c.add_ref(straight_taper())
    ref.dmovey(ii*10)
# c.flatten()
c.show()
c.write_gds("test.gds")
```

**Reply by sebastian-goeldi** [reply]
@JinchW you must decorate your `straight_taper` function with `@gf.cell`, otherwise it will create multiple cells with the same name. You could theoretically also use any other caching decorator such as `@functools.lru_cache(None)`, but they do not allow hashing of lists etc plus do not check the cells as the `@cell` decorator does.

**Reply by JinchW** [reply]
Forgive my negligence. Now this problem seems to no longer exist. Thansks!

**Comment by JinchW** [comment]
Your judgment is correct. I encountered the same problem as you. When the component uses a custom cross_section , the loop program will report an RuntimeError.

**Comment by joamatab** [comment]
Hi, 

could you share the code to reproduce the issue?

we could take a look at it

**Reply by duarte-jfs** [reply]
Hi, the issue I'm facing I believe it's replicated within the code snippet I posted on Sebastian's reply. The issue is also seen when I use all standard gdsfactory. The `taper` function I use is simply an extrusion between two cross sections along a path, so nothing very fancy. But the issue remains even with the straight component with generic pdk cross section. On the code snippet above I expected to have True always.

**Comment by joamatab** [comment]
I also think i saw the same issue when running 

https://github.com/gdsfactory/ubc/pull/396

`samples/test_ubc4.py`

**Reply by sebastian-goeldi** [reply]
Make a breakpoint in the taper function if `gf.kcl.future_cellname == taper_W2_W0p1_TL150_CSF_ffc61053` (or whatever your duplicate cellname function is.

I am pretty sure the problem is that it's a kwargs issue where you are setting a kwarg which gets dropped silently.

**Comment by joamatab** [comment]
i think the problem is that there are 2 functions being used with the same name
